### DISCUSSION TITLE
XIAO nRF52840(sense) With Zephyr(RTOS) | Seeed Studio Wiki

### DISCUSSION THREAD

**Discussion by giscus** [Seeed-Studio/wiki-documents#707] [Q&A]
# XIAO nRF52840(sense) With Zephyr(RTOS) | Seeed Studio Wiki

XIAO nRF52840(sense) With Zephyr(RTOS)

https://wiki.seeedstudio.com/XIAO-nRF52840-Zephyr-RTOS/

**Comment by mandaawad-connected** [comment]
Where to get the zephyr.uf2 file? I was not able to find in the github zephyr project link, thanks!

**Comment by Tjoms99** [comment]
If youre looking for a plug and play zephyr proect there is a battery management lib i haver created on github for the XIAO BLE board: 
https://github.com/Tjoms99/xiao_sense_nrf52840_battery_lib

Here you can read the battery voltage and charge it using the usb cable, enjoy! Feel free to leave feedback

**Reply by MatthewJeffson** [reply]
WoW! Thank you so much for providing it!
Best Regards,
Matthew

**Reply by MatthewJeffson** [reply]
By the way, we just launched open roadmap for XIAO series, and welcome your feedback! https://github.com/Seeed-Studio/XIAO_Series/discussions

**Comment by KiranShingote** [comment]
I am using XIAO_52840 (not sense) board and using nrfutil runner for flashing. After flashing command, I get below error.

-- west flash: using runner uf2
WARNING: runners.uf2: Discovered UF2 partitions don't match Board-ID 'Seeed_XIAO_nRF52840_Sense'
FATAL ERROR: No matching UF2 partitions found

Please guide how to resolve the issue.

**Reply by nfs0619** [reply]
Thank you for your feedback. We have received your message and are currently arranging for debugging. If you do not receive further updates, please feel free to reach out to our technical support at [techsupport@seeed.io](mailto:techsupport@seeed.io) for further assistance. Thank you for your support and understanding.

**Reply by pavel-perina** [reply]
Weird. Mine came with Xiao Sense firmware and it worked until I accidentally overwritten internal flash and had to reupload firmware using Raspberry Pi Pico and SWD interface. I solved this problem by uploading Sense firmware again. Board-ID is hardcoded somewhere in Zephyr's board definition file. I'm not sure how much firmwares differ with exception of USB ID and few strings. I already posted this to nRF wiki

**Comment by dertobias** [comment]
Got a working Matter contact sensor running on the XIAO nRF52840 with NCS v3.2.4, using the stock Adafruit UF2 bootloader — no J-Link, no SWD probe, no bootloader replacement needed.
The key challenge is that the Adafruit bootloader jumps to 0x27000, which conflicts with the standard NCS/MCUboot partition layout. Solved with a custom static partition map and a two-step UF2 flash procedure.

Source code and full write-up including partition layout, gotchas and power optimization:
https://github.com/dertobias/matter-contact-sensor-xiao-nrf52840
